### PR TITLE
Inject real size of interstitial at display time

### PR DESCRIPTION
**Pull request by TurpIF**
JIRA: EE-828

### CODE REVIEW

File: CriteoPublisherSdk/Sources/Display/CR_DisplaySizeInjection.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+// limitations under the License.
+//
+
+#import <Foundation/Foundation.h>
+
+@class CR_DeviceInfo;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface CR_DisplaySizeInjection : NSObject
```
Comment (vguerci) [inline CriteoPublisherSdk/Sources/Display/CR_DisplaySizeInjection.h:26]: Should be named `CR_DisplaySizeInjector`

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h [hunk 2]
```diff
@@ -42,6 +42,12 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (CGSize)fullScreenSize;
 
+/**
+ * Returns the size of the main screen when considering the safe area.
+ * If there is no safe area, then this returns the same than `fullScreeSize`
+ */
+- (CGSize)safeScreenSize;
```
Comment (vguerci) [inline CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h:49]: This is unrelated to commit, which should be atomic

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.m [hunk 3]
```diff
@@ -129,12 +130,12 @@ - (NSString *)deviceId {
   return _deviceId;
 }
 
-+ (CGSize)getScreenSize {
+- (CGSize)fullScreenSize {
```
Comment (vguerci) [inline CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.m:133]: OPT: Now this become an instance method, it could be declared as a property and used with dot notation

File: CriteoPublisherSdk/Sources/CR_HeaderBidding.m [hunk 4]
```diff
@@ -198,6 +202,14 @@ - (void)addCriteoBidToMopubRequest:(id)adRequest
     }
 
     if ([targeting isKindOfClass:[NSString class]]) {
+      NSString *displayUrl = bid.mopubCompatibleDisplayUrl;
+
+      // MoPub interstitial restrains itself to the safe area.
+      if (adUnit.adUnitType == CRAdUnitTypeInterstitial) {
```
Comment (vguerci) [inline CriteoPublisherSdk/Sources/CR_HeaderBidding.m:208]: Unless mistaking this codepath isn't tested, is it?

File: CriteoPublisherSdk/Sources/Display/CR_DisplaySizeInjection.h (hunk 1 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/Display/CR_DisplaySizeInjection.h:26]: Done

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.m (hunk 3 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.m:133]: Done

File: CriteoPublisherSdk/Sources/CR_HeaderBidding.m (hunk 4 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/CR_HeaderBidding.m:208]: Yes it is there: https://github.com/criteo/ios-publisher-sdk/blob/586fd738b2e5a7144b8edde2bef62b8d0155648b/CriteoPublisherSdk/Tests/UnitTests/CR_HeaderBiddingTests.m#L235

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h (hunk 2 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h]: The injector is using this to get the safe area size. Don't you find that it is related ?

Comment (vguerci) [reply CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h]: The commit is worded about injector, not the safe area retrieval. I personally would have split the commits, but that is NIT ¯\_(ツ)_/¯

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h [hunk 5]
```diff
@@ -33,11 +33,16 @@ NS_ASSUME_NONNULL_BEGIN
...
+ */
+@property(nonatomic, readonly) CGSize screenSize;
+
 - (instancetype)initWithThreadManager:(CR_ThreadManager *)threadManager NS_DESIGNATED_INITIALIZER;
 
 - (void)waitForUserAgent:(void (^_Nullable)(void))completion;
-+ (CGSize)getScreenSize;
-+ (BOOL)validScreenSize:(CGSize)size;
+
+- (BOOL)validScreenSize:(CGSize)size;
```
Comment (vguerci) [inline CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h:51]: Probably unrelated to this PR but this one does not look like properly named. Should be something like `isScreenSizeValid:` or `validateScreenSize:`

File: CriteoPublisherSdk/Sources/CR_HeaderBidding.m (hunk 4 above)
Comment (vguerci) [reply CriteoPublisherSdk/Sources/CR_HeaderBidding.m:208]: I meant the whole branch, are you checking that interstitials get their displayUrl altered?

Comment (TurpIF) [reply CriteoPublisherSdk/Sources/CR_HeaderBidding.m:208]: Yes it is, the injector is stubbed to return a display.url value when getting the bid display URL. Then this display.url value  is checked.

File: CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h (hunk 5 above)
Comment (TurpIF) [reply CriteoPublisherSdk/Sources/Configuration/CR_DeviceInfo.h:51]: I think this does not belong to this PR.

File: CriteoPublisherSdk/Tests/UnitTests/Configuration/CR_DeviceInfoTests.m [hunk 6]
```diff
@@ -112,4 +112,19 @@ - (void)testWebViewReleasedAfterUse {
   [self cr_waitForExpectations:@[ expectation ]];
 }
 
+- (void)testSafeScreenSize_GivenAnyDevice_ReturnNonZeroSizeBelowFullScreenSize {
```
Comment (vguerci) [inline CriteoPublisherSdk/Tests/UnitTests/Configuration/CR_DeviceInfoTests.m:115]: What do you think it about at some point to introduce multiple devices in tests? I think GA does that pretty well...

Comment (TurpIF) [reply CriteoPublisherSdk/Tests/UnitTests/Configuration/CR_DeviceInfoTests.m:115]: Yes that would be great. With different versions too. But this may takes a really long time (or not considering that GA jobs are done in //)

Comment (vguerci) [reply CriteoPublisherSdk/Tests/UnitTests/Configuration/CR_DeviceInfoTests.m:115]: GA can do matrix testing I think... ;p
